extensions/amp-video/0.1/amp-video.js

```diff
@@ -204,38 +243,131 @@ class AmpVideo extends AMP.BaseElement {
+    this.propagateAttributes(ATTRS_TO_PROPAGATE_ON_LAYOUT,
+        dev().assertElement(this.video_),
         /* opt_removeMissingAttrs */ true);
 
-    this.getRealChildNodes().forEach(child => {
-      // Skip the video we already added to the element.
-      if (this.video_ === child) {
-        return;
-      }
-      if (child.getAttribute && child.getAttribute('src')) {
-        assertHttpsUrl(child.getAttribute('src'),
-            dev().assertElement(child));
+    if (viewer.getVisibilityState() == VisibilityState.PRERENDER) {
+      if (!this.element.hasAttribute('preload')) {
+        this.video_.setAttribute('preload', 'auto');
       }
-      this.video_.appendChild(child);
-    });
+      this.propagatePrerenderSources_();
+      viewer.whenFirstVisible().then(() => {
+        this.propagateLayoutChildren_();
+      });
+    } else {
+      this.propagatePrerenderSources_();
+      this.propagateLayoutChildren_();
+    }
 
     // loadPromise for media elements listens to `loadstart`
     return this.loadPromise(this.video_).then(() => {
       this.element.dispatchCustomEvent(VideoEvents.LOAD);
     });
   }
 
+  /**
+   * @private
+   * Propagate sources that are cached by the CDN.
+   */
+  propagatePrerenderSources_() {
+    dev().assert(this.video_);
+
+    const sources = toArray(childElementsByTag(this.element, 'source'));
+
+    // if the `src` of `amp-video` itself is cached, move it to <source>
+    if (this.element.hasAttribute('src') && this.isCachedByCDN_(this.element)) {
+      const src = this.element.getAttribute('src');
+      const ampOrigSrc = this.element.getAttribute('amp-orig-src');
+      assertHttpsUrl(src, this.element);
+      const srcSource = this.element.ownerDocument.createElement('source');
+      srcSource.setAttribute('src', src);
+      srcSource.setAttribute('amp-orig-src', ampOrigSrc);
+      sources.unshift(srcSource);
+    }
+
+    sources.forEach(source => {
+      if (this.isCachedByCDN_(source)) {
+        this.video_.appendChild(source);
+      }
+    });
+  }
+
+  /**
+   * Propagate origin sources and tracks
+   * @private
+   */
+  propagateLayoutChildren_() {
+    dev().assert(this.video_);
+
+    const sources = toArray(childElementsByTag(this.element, 'source'));
+
+    // If the `src` of `amp-video` itself is NOT cached, set it on video
+    if (this.element.hasAttribute('src') &&
+        !this.isCachedByCDN_(this.element)) {
+      assertHttpsUrl(this.element.getAttribute('src'), this.element);
+      this.propagateAttributes(['src'], dev().assertElement(this.video_));
+    }
+
+    sources.forEach(source => {
+      // Cached sources should have been moved already.
+      dev().assert(!this.isCachedByCDN_(source));
+      assertHttpsUrl(source.getAttribute('src'), source);
+      this.video_.appendChild(source);
+    });
+
+    // To handle cases where cached source may 404 if not primed yet,
+    // duplicate the `origin` Urls for cached sources and insert them after each
```

Comment:
Does the browser automatically proceed to the next source if one fails? If that is true we should comment it here.

Reply:
luckily yes, updated comments